### PR TITLE
Fixed icarl by converting back TrainEvalModel to DynamicModule

### PR DESCRIPTION
**Pull request by AndreaCossu**
The behavior of TrainEvalModel was not always consistent and the classifier sometimes failed to switch.
I brought back the DynamicModule version, this fixes iCarl performance.

### CODE REVIEW

**Comment by AntonioCarta** [comment]
This is still broken though. You shouldn't have to adapt again on already seen experiences.

**Comment by AndreaCossu** [comment]
I agree, I am working on fixing this but for the moment I just wanted to make iCARL work. I need to further inspect why the flag `self.training` was not switching properly.

**Comment by AndreaCossu** [comment]
The last commits should have fixed the missing switch between train and eval in iCarl. The behavior is pretty strange. Basically, before the change the model started in training mode, then it switched to eval mode after the first 'strategy.eval' call and then it stayed in eval forever.
I had to remove both the 'old_model.eval()' call in the icarl loss and add the 'strategy.model.train()' in the icarl plugin. 

This fixes the switch but it is suspicious, the model should automatically switch back to training mode each time 'strategy.train' is called.

**Comment by AntonioCarta** [comment]
> the model should automatically switch back to training mode each time 'strategy.train' is called.

It's a bit different, `strategy.eval` stores the mode for each module and reset them at the end of the method.
